sql/2021/resource-hints/dns-prefetch_domains.sql

```diff
@@ -0,0 +1,65 @@
+#standardSQL
+# Most popular domains users dns-prefetch from
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+CREATE TEMPORARY FUNCTION getResourceHints(payload STRING, hint STRING)
+RETURNS STRING
+LANGUAGE js AS '''
+try {
+  var $ = JSON.parse(payload);
+  var almanac = JSON.parse($._almanac);
+  return JSON.stringify(almanac['link-nodes'].nodes.filter(link => link.rel.toLowerCase() == hint));
+} catch (e) {  
+  return "";
+}
+''' ;
+
+CREATE TEMPORARY FUNCTION getHrefs(payload STRING)
+RETURNS ARRAY<STRING>
+LANGUAGE js AS '''
+try {
+    var $ = JSON.parse(payload);
+    return $.map(n => n.href);
+} catch (e) {
+    return [];
+}
+''' ;
+
+SELECT
+  client,
+  domain,
+  COUNT(0) AS freq,
+  AS_PERCENT(COUNT(0), SUM(COUNT(0)) OVER()) AS pct
+FROM (
+    SELECT
+      client,
+      REGEXP_EXTRACT(href, r'\/\/(.+?)$') AS domain
+    FROM (
+        SELECT
+          client,
+          getHrefs(hints) AS value
+        FROM (
+            SELECT
+              _TABLE_SUFFIX AS client,
+              getResourceHints(payload, "dns-prefetch") AS hints
+            FROM
+              `httparchive.pages.2021_07_01_*`
+        )
+    )
+    CROSS JOIN UNNEST(value) AS href
```

Comment:
With the functions combined, this section could be consolidated
```suggestion
    SELECT
      _TABLE_SUFFIX AS client,
      REGEXP_EXTRACT(href, r'\/\/(.+?)$') AS domain
    FROM
      `httparchive.pages.2021_07_01_*`,
      UNNEST(getResourceHintHrefs(payload, "dns-prefetch") AS href
```

Reply:
Don't think I could consolidate as much as if I consolidate the three SELECT statements I would be unable to use the condition
```
  WHERE
    ARRAY_LENGTH(value) > 0
```

and need to replace it with 

```
  WHERE
    ARRAY_LENGTH(getResourceHintHrefs(payload, "dns-prefetch")) > 0
```

I can however consolidate two of the statements.

Reply:
You could probably remove that part from the query and it'd still work because the CROSS JOIN UNNEST won't produce any `href` values for pages in which the array is empty.